### DISCUSSION TITLE
How to dispose resource in custom AbstractUIExtension

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-glsp/glsp#999] [Q&A] [answered]
Hi,
    I create a custom ui extension in my glsp diagram widget. I have a setInterval in my custom control. so I want to dispose of it when the diagram gets closed, I am not sure what the better solution is, Could anyone give some ideas? thank you very much

**Comment by tortmayr** [comment, marked as answer]
HI @wss29,
I think the easiest approach to achieve this is configuring your own custom `GLSPDiagramWidget` and customizing the `onAfterDetach` method e.g. like this:
```ts
export class MyDiagramWidget extends GLSPDiagramWidget {
    protected override onAfterDetach(msg: Message): void {
       const extension= this.diContainer.get<MyCustomUIExtension>(MyCustomUIExtension);
       extension.dispose();
    }
}
```

In addition, you then need to bind a custom `DiagramManager`:
```ts
@injectable()
export class MyDiagramManager extends GLSPDiagramManager {
    //...

    override async createWidget(options?: any): Promise<GLSPDiagramWidget> {
        if (GLSPDiagramWidgetOptions.is(options)) {
            const clientId = this.createClientId();
            const widgetId = this.createWidgetId(options);
            const config = this.getDiagramConfiguration(options);
            const diContainer = config.createContainer(clientId);

            // do not await the result here as it blocks the Theia layout restoration for open widgets
            // instead simply check in the widget if we are already initialized
            this.diagramConnector.initializeResult.then(initializeResult =>
                configureServerActions(initializeResult, this.diagramType, diContainer)
            );

            const widget = new MyDiagramWidget(
                options,
                widgetId,
                diContainer,
                this.editorPreferences,
                this.storage,
                this.theiaSelectionService,
                this.diagramConnector
            );
            widget.listenToFocusState(this.shell);
            return widget;
        }
        throw Error('DiagramWidgetFactory needs DiagramWidgetOptions but got ' + JSON.stringify(options));
    }
}
```
(We are areware that customizing the diagram widget is currently a bit clunky and this will be improved with the upcoming release, but for now this is the way to do it)

**Reply by wss29** [reply]
Hi,
   Thank you very much, what is the time of the next release, I think there will be a lot of updates, looking forward to the new release

**Reply by wss29** [reply]
Hi,
   Do I need to add a bind to MyDiagramManager in the front module or somewhere else
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/e00968da-4a55-4f26-aec3-7187a1f7e7b9)
I have seen these code, I am not sure if my custom has some effect, or should MyDiagramManager extends ConfigurableGLSPDiagramManager?
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/e81c0fac-1771-4cfe-bf3c-b1ff39c9aaf3)

**Reply by tortmayr** [reply]
> Hi, Thank you very much, what is the time of the next release, I think there will be a lot of updates, looking forward to the new release

The release is scheduled for end June/ beginning of July 2023. In the meantime, if you want to have a peek at the upcoming version you can check out the monthly release candidates (currently version: 1.1.0-RC08)

>Do I need to add a bind to MyDiagramManager in the front module or somewhere else

You need to override the `configureDiagramManager` method in your frontend module:

```ts
configureDiagramManager(context: ContainerContext): void {
   registerDiagramManager(context, YourCustomDiagramManager)
}
```

>or should MyDiagramManager extends ConfigurableGLSPDiagramManager?
No your custom manager should not extend from the `ConfigurableGLSPDiagramManager` class. This is an internal class.

**Reply by wss29** [reply]
Thanks a great
